lib/web/device_trust.go

```diff
@@ -76,7 +81,44 @@ func (h *Handler) deviceWebConfirm(w http.ResponseWriter, r *http.Request, _ htt
 
 	// Always redirect back to the dashboard, regardless of outcome.
 	app.SetRedirectPageHeaders(w.Header(), "" /* nonce */)
-	http.Redirect(w, r, "/web", http.StatusSeeOther)
+
+	redirectTo, err := h.getRedirectURL(redirectURI)
+	if err != nil {
+		h.log.
+			WithError(err).
+			WithField("redirect_uri", redirectURI).
+			Debug("Unable to parse redirectURI")
+	}
+	http.Redirect(w, r, redirectTo, http.StatusSeeOther)
 
 	return nil, nil
 }
+
+// getRedirectURL tries to parse the given redirectURI. It will always return a redirect url
+// even if the parse fails (in case of failture, the returned string is "/web")
+func (h *Handler) getRedirectURL(redirectURI string) (string, error) {
+	const basePath = "/web"
+
+	if redirectURI == "" {
+		return basePath, nil
+	}
+
+	parsedURL, err := url.Parse(redirectURI)
+	if err != nil {
+		return basePath, trace.Wrap(err)
+	}
+
+	cleanPath := path.Clean(parsedURL.Path)
+	// helps in situations where there is no path such as https://example.com
+	if cleanPath == "." || cleanPath == ".." {
+		cleanPath = "/"
+	} else if !strings.HasPrefix(cleanPath, "/") {
```

Comment:
In what situation is `cleanPath` not equal to `.` or `..` and  also doesn't have the `/` prefix?

https://go.dev/play/p/bOEv7v_qAb5

Reply:
SSO redirect may return something like `example/path` (I haven't seen it in actual testing but the fact that that param _could_ exist caused me to add this check).

The client redirect URLs from SSO callbacks generally _don't_ have the hostname attached and are only the paths, so I tried to catch it all (and added tests for them)